### PR TITLE
Add dfdlx:runtimeProperties extension

### PR DESCRIPTION
**Pull request by stevedlawrence**
New dfdlx:runtimeProperties is a space-separated list of key=value
pairs. The pairs are converted to a Map and made available on the
ElementRuntimeData for which the properties is specified. This Map can
be accessed by custom implementations of InfosetInputters and
InfosetOutputters to allow modifications and custom behaviors to the
text of simple types.

The property is only valid on dfdl:element and xs:element tags for
simple types.

Also changed "Object" to "ObjectKind" to make it more clear that it
describes the dfdlx:objectKind extension property.

DAFFODIL-2537

### CODE REVIEW

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetInputter.scala [hunk 1]
```diff
@@ -121,6 +121,13 @@ abstract class InfosetInputter
...
    * simple content, this should return either null or the empty string.
    */
+  def getSimpleText(primType: NodeInfo.Kind, runtimeProperties: Map[String,String]): String =
+    getSimpleText(primType)
+
+  /**
+   * See getSimpleText(primType, runtimeProperties), which has a default
+   * implementation to call this function without the runtimeProperties Map
+   */
   def getSimpleText(primType: NodeInfo.Kind): String
```
Comment (tuxji) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetInputter.scala:131]: I question the need to overload `getSimpleText` with one and two parameters.  Wouldn't it be easier to add a default argument to getSimpleText instead?

```scala
def getSimpleText(primType: NodeInfo.Kind, runtimeProperties: Map[String,String] = Map.empty[String, String]): String
```

Passing `Map.empty` as a default argument removes the need to worry about handling `null` or `Option` or auto-converting `Map` to `Option[Map]` (see StackOverflow on [optional parameters](https://stackoverflow.com/questions/4199393/are-options-and-named-default-arguments-like-oil-and-water-in-a-scala-api)).

The corresponding Java API still would need to overload `getSimpleText` with one and two parameters since Java doesn't have default parameters, but you still could simplify the corresponding Scala API the same way.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/RuntimeData.scala [hunk 2]
```diff
@@ -697,7 +698,8 @@ sealed abstract class ErrorERD(local: String, namespaceURI: String)
     null, // fillByteEvArg => FillByteEv
     Nope, // maybeCheckByteAndBitOrderEvArg: => Maybe[CheckByteAndBitOrderEv],
     Nope, // maybeCheckBitOrderAndCharsetEvArg: => Maybe[CheckBitOrderAndCharsetEv],
-    false // isQuasiElementArg: => Boolean
+    false, // isQuasiElementArg: => Boolean
+    null, // runtimeProperties: Map[String,String]
```
Comment (tuxji) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/RuntimeData.scala:702]: Even though this is only an `ErrorERD`, we might be safer passing `Map.empty` rather than null so that an `InfosetInputter ` doesn't end up calling `getSimpleText` with null as the second argument.

**Review by tuxji — APPROVED** [review]
+1 with some questions to discuss...

Is there any asciidoc design document as mentioned in DAFFODIL-2537?  Is it worth adding an implementation and a test showing the motivating example of a custom XML Infoset Inputter and Outputter for embedded XML strings in binary data?

See review comments for additional questions.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetInputter.scala (hunk 1 above)
Comment (stevedlawrence) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetInputter.scala:131]: I was trying to prevent any issues with backwards compatibility.

Thinking about it though, this InfosetInputter.scala stuff is "internal" API, so I think it shouldn't be a problem to have this use a default parameter. In fact, since this is internal, this really can drop the single-arg version entirely--Daffodil internals will never call this single-arg version. This does require updating all of our built-in infoset inputters to use the two-arg function, but that's trivial to just add the new parameter.

But I think the public Java/Scala API infoset inputters still need the changes I've added to Infoset.scala. Otherwise, we'll break existing InfosetInputters that don't implement the two arg version. I think...

Note that we also need to take a look at our InfosetOutputter/Inputter API's (see [Daffodil1909](https://issues.apache.org/jira/browse/DAFFODIL-1909)), so we'll probably need to break things at some point. So maybe we break things for that change?

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/ElementBase.scala [hunk 3]
```diff
@@ -1179,4 +1179,17 @@ trait ElementBase
         Nil
     res
   }
+
+  lazy val runtimeProperties = findPropertyOption("runtimeProperties").toOption match {
```
[the remark this answers is not in the document]
Comment (stevedlawrence) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/ElementBase.scala:1183]: Currently this is specific only to infoset inputters/outputters, but there's potential for other runtimes to use this in other ways. For example, maybe the runtime2 could use this property to enable some specific things that only runtime2 cares about. But I can add some document to that affect. It's up to runtimes to decide when and how they want to use this map.

There is no documentation for this property, I'll add a page to the wiki to describe it.

File: daffodil-japi/src/main/scala/org/apache/daffodil/japi/infoset/Infoset.scala [hunk 4]
```diff
@@ -67,13 +70,29 @@ abstract class InfosetInputter extends SInfosetInputter {
...
   def getNamespaceURI(): String
 
+  /**
+   * Wrapper function to convert the runtimeProperties Scala Map to a Java Map
+   * and call the getSimpleText function. Proably should not be overridden by
+   * those using the Java API. Instead override getSimpleText(primType, runtimeProperties)
+   * or getSimpleText(primType)
+   */
+  override def getSimpleText(primType: NodeInfo.Kind, runtimeProperties: Map[String,String]): String =
+    getSimpleText(primType, runtimeProperties.asJava)
```
[the remark this answers is not in the document]
Comment (stevedlawrence) [reply daffodil-japi/src/main/scala/org/apache/daffodil/japi/infoset/Infoset.scala:80]: Yeah, I noticed that :( In order to cover this we would need to write a custom infoset inputter in Java (and probably also in Scala for the Scala API). It's doable, but probably a pain. I'll see if I can come up with a real simple infoset inputter that avoids a lot of complexity that infoset inputters/outputters normally have.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/RuntimeData.scala (hunk 2 above)
Comment (stevedlawrence) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/RuntimeData.scala:702]: I almost wonder if this is a good thing error? Fields in this ErrorERD should never be used, so having null values would ensure if they were used we could get a NPE and it would be very obvious something was using this when it shouldn't be. If we have valid empty values, we might not realize that some code is using it.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetInputter.scala (hunk 1 above)
Comment (tuxji) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetInputter.scala:131]: I'm fine with changing the internal InfosetInputter API and all the calls to use two arguments (no default parameter needed or used), although that's more work for you.  For backwards compatibility, I agree the public Java/Scala API InfosetInputters still need to use overloaded one-arg and two-arg functions (Java) or a default parameter (Scala), although a default parameter still breaks binary compatibility but not source compatibility (if that matters).  

Did you see my questions about the asciidoc design document and the custom XML InfosetInputter/Outputter for embedded XML strings in binary data too?

File: daffodil-core/src/test/scala/org/apache/daffodil/general/TestRuntimeProperties.scala [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+import org.apache.daffodil.infoset.DISimple
+import org.apache.daffodil.infoset.ScalaXMLInfosetInputter
+import org.apache.daffodil.infoset.ScalaXMLInfosetOutputter
+import org.apache.daffodil.io.InputSourceDataInputStream
+import org.apache.daffodil.util.SchemaUtils
+import org.apache.daffodil.util.TestUtils
+
+/**
+ * Test InfosetOutputter that extends the ScalaXMLInfosetOutputter and redacts
+ * words specified in the dfdlx:runtimeProperties extension
```
Comment (mbeckerle) [inline daffodil-core/src/test/scala/org/apache/daffodil/general/TestRuntimeProperties.scala:36]: This redaction example is perhaps too simple. 

An interesting thing is when the special infoset outputter creates XML that is NOT valid w.r.t. the DFDL schema interpreted as an XML Schema.  One then needs a separate schema for validation from the one used for parsing/unparsing. 

If you add a pattern facet to the filtered element below in your DFDL schemas, such that things with the word REDACTED in them will NOT be valid w.r.t the pattern, that would be more representative of how this feature can change the XML in ways that violate the XML validity. 

This is tricky with respect to Daffodil's validation behaviors. "limited" validation will validate still, as it sees the infoset data during parsing. "full" validation should take the output of the infoset outputter and feed it to Xerces for validation, and that should create validation errors. 

I think tests which characterize these two different interactions of this feature with validation would be good additions.

File: daffodil-core/src/test/scala/org/apache/daffodil/general/TestRuntimeProperties.scala [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+class TestRuntimeProperties {
+
+  val testSchema = SchemaUtils.dfdlTestSchema(
+    <xs:include schemaLocation="org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd"/>,
+    <dfdl:format ref="tns:GeneralFormat" lengthKind="delimited"/>,
+    <xs:element name="root">
+      <xs:complexType>
+        <xs:sequence dfdl:separator=",">
+          <xs:element name="filtered" type="xs:string" dfdlx:runtimeProperties="redact=foo,baz redactWord=REDACTED" />
```
Comment (mbeckerle) [inline daffodil-core/src/test/scala/org/apache/daffodil/general/TestRuntimeProperties.scala:91]: Per suggestion above, make another test schema which is just like this one, but which adds a pattern facet to filtered disallowing the "REDACTED" from showing up in the data. So that if redaction occurs, the resulting data will be invalid. Then have tests detect this validation error, but only if "full" validation is on.

**Review by mbeckerle — CHANGES_REQUESTED** [review]

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/RuntimeData.scala (hunk 2 above)
Comment (mbeckerle) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/RuntimeData.scala:702]: Yeah, I guess I'll go with null is ok here as well.

File: daffodil-propgen/src/main/resources/org/apache/daffodil/xsd/dfdlx.xsd [hunk 7]
```diff
@@ -144,4 +145,16 @@
...
+  <xs:attributeGroup name="RuntimePropertiesAG">
+    <xs:attribute form="qualified" name="runtimeProperties" type="dfdlx:RuntimePropertiesType" />
+  </xs:attributeGroup>
+
+  <xs:attributeGroup name="RuntimePropertiesAGQualified">
+    <xs:attribute form="qualified" name="runtimeProperties" type="dfdlx:RuntimePropertiesType" />
+  </xs:attributeGroup>
+
+  <xs:simpleType name="RuntimePropertiesType">
+    <xs:restriction base="xs:string" />
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-propgen/src/main/resources/org/apache/daffodil/xsd/dfdlx.xsd:157]: To follow up on this; however, a list type can be separated by any whitespace, your parsing code looks for single spaces. Only attributes get their values normalized to single spaces. While infrequently used, we do have this Element-form syntax like:
```
<dfdl:property name="dfdlx:runtimeProperties">key=value1
key=value2
key=value3
</dfdl:property>
```
And in that situation the list would not be separated by individual spaces, but by whitespace generally, possibly more than 1 character. 
You can park this as a separate JIRA ticket that dfdlx:runtimeProperties is incompatible with dfdl:property style annotations.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/RuntimeData.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/RuntimeData.scala:702]: null is ok here per reasoning in earlier comments in this review.

File: daffodil-propgen/src/main/resources/org/apache/daffodil/xsd/dfdlx.xsd (hunk 7 above)
[the remark this answers is not in the document]
Comment (stevedlawrence) [reply daffodil-propgen/src/main/resources/org/apache/daffodil/xsd/dfdlx.xsd:157]: I think I can just change the split to happen on `\s+` instead of single-space to support that. I'd rather add that now while this is all fresh in my mind.

File: daffodil-core/src/test/scala/org/apache/daffodil/general/TestRuntimeProperties.scala (hunk 5 above)
Comment (stevedlawrence) [reply daffodil-core/src/test/scala/org/apache/daffodil/general/TestRuntimeProperties.scala:36]: This raises an interesting issue/limitation.

The way Daffodil does full validation is it tee's the infoset events to two InfosetOutputters: one provided by the user (e.g. our redacting outputter) and the other a normal XMLTextInfosetOutputter. At the end of parsing, we run Xerces validation on the result of the XMLTextInfosetOutputter. This is necessary because Daffodil only knows how to validate the result of the XMLTextInfosetOutputter. If the user provided InfosetOutputter output json, NiFi records, some proprietary thing, etc. it wouldn't be able enable full validation otherwise.

And the runtimeProperty redaction/transformation only affects the output of the user provied InfosetOutputter. It does not modify the internal infoset, and it does not propagate any changes to the XMLTextInfosetOutputter used for full validation.

So Daffodil validation has no knowledge about what changes the user provided InfosetOutputter makes, and so enabling full validation change anything. Daffodil only validates the infoset as parsed.

We can think of InfosetOutputters as post-processing of the infoset, and validation occurs before that post-processing takes place. If something is expecting to validate the post-processed data make by a custom InfosetOutputter, it needs to do that in a different step outside of Daffodil.

Comment (mbeckerle) [reply daffodil-core/src/test/scala/org/apache/daffodil/general/TestRuntimeProperties.scala:36]: Ah, so in that case there's no need for the tests that address any interactions with validation, as there is none.

**Comment by stevedlawrence** [comment]
I've pushed updates that I think should address the comments, this includes new test InfosetInputters using the Scala/Java API to ensure coverage. It also shows some issues. For example, the Java API needs to access some Scala objects, which is pretty awkward.

>  Is it worth adding an implementation and a test showing the motivating example of a custom XML Infoset Inputter and Outputter for embedded XML strings in binary data?

I don't really want to add such an implementation, just because I suspect it's going to be a decent amount of work, and a decent amount of testing. I think it's basically a implementation of one of the InfosetInputters/Outputters with quite a bit of extra complicated logic. If we do want to include that in the repo I think that's worth done in another PR. Once I add some documentation (next week) I'll add this as a motivating example and probably include some pseuocode about how this might be implemented.

**Review by mbeckerle — APPROVED** [review]
+1
Looks good. Adding a wiki page about this would be good as we need such for each extension to DFDL. 
Writing an infoset inputter/outputter that illustrate dealing with strings of XML (or maybe JSON) inside binary data is needed, but I agree it can go in a separate ticket/PR.

File: daffodil-japi/src/main/scala/org/apache/daffodil/japi/infoset/Infoset.scala (hunk 4 above)
Comment (stevedlawrence) [inline daffodil-japi/src/main/scala/org/apache/daffodil/japi/infoset/Infoset.scala:80]: I was wondering if it's worth getting rid of this `asJava`? Looking at the implementation, asJava allocates a new `MapWrapper`, which implements the `java.util.AbstractMap` interface to access the underlying Scala `Map`. So by using `asJava` we at least aren't creating a completely new `Map` and whatever allocations are involved with that, but we are still allocating this wrapper class. And we do this even when this Maps are empty (which is going to be the most common case). And even worse, most implementations of this Infoset are not even going to implement this two arg version of getSimpleText, so they'll this allocation will just be unused.

So, I'm wondering if when we create the `val runtimeProperties` in ElementBase.scala, we should just create a Java Map, which scala API users can use without a problem, and we never need to allocate anything extra? We could use the `asJava` method and create the wrapper mentioned above, or we could directly create a JavaMap to avoid that extra allocation and indirection? Any thoughts/concerns with this?

Comment (tuxji) [reply daffodil-japi/src/main/scala/org/apache/daffodil/japi/infoset/Infoset.scala:80]: `asJava`'s new `MapWrapper` allocation occurs only if a Java API user calls this specific two-arg `getSimpleText` method, correct?  There's a third choice worth adding to the previous two choices then:

1. Create Java Map instead of Scala Map in ElementBase.scala.
2. Continue using Scala Map, calling `asJava` and creating wrapper in this method.
3. Check for empty Scala Map in this method and if so, call Java Collections.emptyMap() instead of calling `asJava` and creating wrapper.  This choice is more efficient because it doesn't create a new wrapper object; it just re-uses an existing empty and immutable object.

I think the third choice might be the best compromise.

Comment (tuxji) [reply daffodil-japi/src/main/scala/org/apache/daffodil/japi/infoset/Infoset.scala:80]: OK, a Java Map in ElementBase sounds good to me then.

**Comment by stevedlawrence** [comment]
A writeup of these feature, including a potential implementation using Scala XML Nodes for simplicity and brevity is added here:

https://cwiki.apache.org/confluence/display/DAFFODIL/Proposal%3A+Runtime+Properties
